torch/_ops.py

```diff
@@ -398,50 +397,109 @@ def key_extractor(tensors, key_mask):
+    current_mode_stack_pre_dispatch = mode_stack_state_for_pre_dispatch()
+    assert mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    if mode_key == torch._C._TorchDispatchModeKey.PROXY:
+        current_mode = current_mode_stack_pre_dispatch.get(0)
+        mode_stack_state_for_pre_dispatch().set(0, None)
+        return current_mode
+    else:
+        current_mode = current_mode_stack_pre_dispatch.get(1)
+        mode_stack_state_for_pre_dispatch().set(1, None)
+        return current_mode
 
-# Per-dispatch-key mode variant.
-# Temporarily pops the top of a given mode stack.
-@contextlib.contextmanager
-def temporarily_pop_mode(mode_stack):
-    assert len(mode_stack) > 0
-    top_mode = mode_stack.pop()
-    try:
-        yield top_mode
-    finally:
-        mode_stack.append(top_mode)
 
+def _set_mode_pre_dispatch(mode):
+    from torch._subclasses.functional_tensor import FunctionalTensorMode
+    from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode
+
+    assert isinstance(mode, (FunctionalTensorMode, ProxyTorchDispatchMode))
+    if isinstance(mode, FunctionalTensorMode):
+        current_mode = mode_stack_state_for_pre_dispatch().get(1)
+        assert current_mode is None
+        mode_stack_state_for_pre_dispatch().set(1, mode)
+        return
+
+    current_mode = mode_stack_state_for_pre_dispatch().get(0)
+    assert current_mode is None
+    mode_stack_state_for_pre_dispatch().set(0, mode)
+
+
+def _pop_mode_from_pre_dispatch():
+    mode_stack = mode_stack_state_for_pre_dispatch()
+    if mode_stack.get(1) is not None:
+        res = mode_stack.get(1)
+        mode_stack.set(1, None)
+        return res
+
+    if mode_stack.get(0) is not None:
+        res = mode_stack.get(0)
+        mode_stack.set(0, None)
+        return res
+
+    raise AssertionError("Trying to pop empty mode stack")
+
+
+def _len_torch_dispatch_stack_pre_dispatch():
+    return mode_stack_state_for_pre_dispatch().count()
 
-def mode_stack_per_key():
-    global _mode_stack_per_key
-    return _mode_stack_per_key
 
+def _get_dispatch_mode_pre_dispatch(mode_key):
+    assert mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    if mode_key == torch._C._TorchDispatchModeKey.PROXY:
+        return mode_stack_state_for_pre_dispatch().get(0)
+    return mode_stack_state_for_pre_dispatch().get(1)
 
-# Per-dispatch-key mode variant of push_mode().
-def push_mode_for_key(key, mode):
-    assert isinstance(key, torch._C.DispatchKey)
-    assert isinstance(mode, torch.utils._python_dispatch.TorchDispatchMode)
-    if key not in mode_stack_per_key():
-        mode_stack_per_key()[key] = []
-    mode_stack_per_key()[key].append(mode)
+
+def _get_current_dispatch_mode_pre_dispatch():
+    stack_len = mode_stack_state_for_pre_dispatch().count()
+    # Return a user mode on the stack if there are any
```

Comment:
not sure what this means - there can't be any user modes on the pre-dispatch stack, right?

Reply:
Lol yeah i was just copying code lol